### PR TITLE
EREGCSC-1952 Admin panel import csv button

### PR DESCRIPTION
**Pull request by thwalker6**
Resolves # EREGCSC1052

**Description-**

Adds an import button for csv of supplemental content

**This pull request changes...**

- Adds an import button to the supplemental content list page
- Adds a page that button directs you to to upload a csv
- Creates a landing page
- Refactors the location checker for abstract resource supplemental content to be used for this
- Adds a place holder function for the open csv
- Adds a unit test for the refactored function

**Steps to manually verify this change...**

1. Go to the admin panel and then to supplemental content.
2. On the top right you should see a button for import resources
3. Click the button and you will be taken to a new url with instructions on how to upload resources.
4. Click the template and open the csv.   make sure it has name, description, url, category and locations on the top row.
5. Use the template and put in the next row :

- content-1
- description-1
- www.fake.com (unique urls are checked in admin so if its a url that already exist it wont create)
- implementation resources
- 42 cfr 400.200, blah

Save and then click the uplaod csv button. 
Select the newly created csv and then click import resources.
It will redirect to a new page that will show the created content, and the added/non added locations.  You should see "blah" as a not created.
click on the name of the content which should redirect you to it in the admin panel

### CODE REVIEW

File: solution/backend/resources/admin.py [hunk 1]
```diff
@@ -398,13 +414,66 @@ def save(self, commit=True):
...
+            try:
+                return render(request, "admin/resources_imported.html", context={
+                    "num_conversions": 1,
+                    "conversions": None,
+                })
+            except ValidationError as e:
+                error = e.message
+            except Exception as e:
+                error = str(e)
+        if "get_template" in request.POST:
```
Comment (peggles2) [inline solution/backend/resources/admin.py:445]: does this do anything?

**Review by peggles2 — CHANGES_REQUESTED** [review]
Looks good. I left a few comments.

File: solution/backend/resources/admin.py [hunk 2]
```diff
@@ -162,6 +165,21 @@ def get_queryset(self, request):
...
+        bulk_adds = []
+        bad_locations = []
+        split_locations = bulk_locations.split(",")
+        print(split_locations)
+        for location in split_locations:
+            a = self.build_location(location.strip(), bulk_title)
+            if a:
+                bulk_adds.append(a)
+                # form.instance.locations.add(a.abstractlocation_ptr)
+                # bulk_adds.append(AbstractLocationPolymorphicSerializer(a).data)
```
[the remark this answers is not in the document]
Comment (thwalker6) [reply solution/backend/resources/admin.py:178]: wasnt ready for review, was still a draft, forgot to mark it as such

File: solution/backend/resources/admin.py (hunk 1 above)
Comment (thwalker6) [reply solution/backend/resources/admin.py:445]: wasnt ready for review, was still a draft, forgot to mark it as such

File: solution/backend/resources/admin.py [hunk 3]
```diff
@@ -175,32 +193,30 @@ def save_related(self, request, form, formsets, change):
...
-            split_locations = bulk_locations.split(",")
-            for location in split_locations:
-                a = self.build_location(location.strip(), bulk_title)
-                if a:
-                    form.instance.locations.add(a.abstractlocation_ptr)
-                    bulk_adds.append(AbstractLocationPolymorphicSerializer(a).data)
-                else:
-                    bad_locations.append(location)
+            bulk_adds, bad_locations = self.get_bulk_locations(bulk_locations, bulk_title)
+            print(bulk_adds)
```
[the remark this answers is not in the document]
Comment (thwalker6) [reply solution/backend/resources/admin.py:201]: wasnt ready for review, was still a draft, forgot to mark it as such

File: solution/backend/resources/admin.py [hunk 4]
```diff
@@ -162,6 +165,21 @@ def get_queryset(self, request):
             Prefetch("category", AbstractCategory.objects.all().select_subclasses()),
         )
 
+    def get_bulk_locations(self, bulk_locations, bulk_title=""):
+        bulk_adds = []
+        bad_locations = []
+        split_locations = bulk_locations.split(",")
+        print(split_locations)
```
[the remark this answers is not in the document]
Comment (thwalker6) [reply solution/backend/resources/admin.py:172]: wasnt ready for review, was still a draft, forgot to mark it as such

File: solution/backend/resources/templates/admin/import_resources.html [hunk 5]
```diff
@@ -0,0 +1,36 @@
+{% extends "admin/base_site.html" %}
+
+{% block content %}
+<form action="." method="GET" enctype="multipart/form-data">
+    {% csrf_token %}
+    <p>The csv is required to have the following fields in order:</p>
```
Comment (cgodwin1) [inline solution/backend/resources/templates/admin/import_resources.html:6]: `csv` to `CSV file`

File: solution/backend/resources/templates/admin/import_resources.html [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+    <table>
+        <th>Name</th>
+        <th>Description</th>
+        <th>Url</th>
+        <th>Category</th>
+        <th>locations</th>
+    </table><br>
+    <p>You can get a copy of the template by clicking get template. </p>
```
Comment (cgodwin1) [inline solution/backend/resources/templates/admin/import_resources.html:14]: Maybe quotes around "get template"

File: solution/backend/resources/templates/admin/import_resources.html [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+    <p><input type="submit" name="get_template" value="Get template"/>
+    <p> CSV's seperate fields by commas.  If you use google sheets or excel to make the csv you dont need to worry but if you use a text editor have the format be like below:</p>
```
Comment (cgodwin1) [inline solution/backend/resources/templates/admin/import_resources.html:16]: Recommend capitalizing Google Sheets and Excel, capitalizing CSV and add apostrophe to "don't"

File: solution/backend/resources/templates/admin/import_resources.html [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+        <code>
+    Name, Description, URL, Category, locations <br>
+    conent-1, description-1, url-1, category-1, "location-1, location-2, location-3"<br><br>
```
Comment (cgodwin1) [inline solution/backend/resources/templates/admin/import_resources.html:19]: `conent` to `content`

File: solution/backend/resources/templates/admin/import_resources.html [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
+        <th>Description</th>
+        <th>Url</th>
+        <th>Category</th>
+        <th>locations</th>
+    </table><br>
+    <p>You can get a copy of the template by clicking get template. </p>
+    <p><input type="submit" name="get_template" value="Get template"/>
+    <p> CSV's seperate fields by commas.  If you use google sheets or excel to make the csv you dont need to worry but if you use a text editor have the format be like below:</p>
+        <code>
+    Name, Description, URL, Category, locations <br>
```
Comment (cgodwin1) [inline solution/backend/resources/templates/admin/import_resources.html:18]: Capitalize `locations` for consistency with the rest

**Review by cgodwin1 — COMMENTED** [review]
Code looks great, just some super nitpicky grammatical things from me, not a big deal and not going to necessarily block approval because of them

File: solution/backend/resources/templates/admin/import_resources.html (hunk 7 above)
Comment (thwalker6) [reply solution/backend/resources/templates/admin/import_resources.html:16]: done

File: solution/backend/resources/templates/admin/import_resources.html (hunk 5 above)
Comment (thwalker6) [reply solution/backend/resources/templates/admin/import_resources.html:6]: done

File: solution/backend/resources/templates/admin/import_resources.html (hunk 6 above)
Comment (thwalker6) [reply solution/backend/resources/templates/admin/import_resources.html:14]: done

File: solution/backend/resources/templates/admin/import_resources.html (hunk 9 above)
Comment (thwalker6) [reply solution/backend/resources/templates/admin/import_resources.html:18]: done

File: solution/backend/resources/templates/admin/import_resources.html (hunk 8 above)
Comment (thwalker6) [reply solution/backend/resources/templates/admin/import_resources.html:19]: done

**Comment by peggles2** [comment]
Tried to upload a blank content upload.csv and it just gave me a blank page
Screen Shot 2023-06-22 at 10 54 28 AM

**Pull request merged**